### PR TITLE
Ensure PFS service started when pf.start_trace

### PR DESCRIPTION
**Pull request by YingChen1996**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_start_trace.py [hunk 1]
```diff
@@ -41,6 +48,36 @@ def start_trace():
     print(f"You can view the trace from UI url: {ui_url}")
 
 
+def _start_pfs_in_background(pfs_port) -> None:
+    """Start a pfs process in background."""
+    if is_port_in_use(pfs_port):
+        _logger.debug("PFS is already serving on port %s", pfs_port)
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/_sdk/_start_trace.py:54]: is it possible that another application uses this port? according to current code, if I start PFS with `--port` parameter, we will NOT record the port in config - maybe we can make port always record after it starts?

File: src/promptflow/promptflow/_sdk/_start_trace.py [hunk 2]
```diff
@@ -41,6 +48,36 @@ def start_trace():
...
+        return
+    args = [sys.executable, "-m", "promptflow._sdk._service.entry", "start"]
+    # Start a pfs process using detach mode
+    if platform.system() == "Windows":
+        os.spawnv(os.P_DETACH, sys.executable, args)
+    else:
+        os.system(" ".join(["nohup"] + args + ["&"]))
+
+
+def _liveness_probe(port, interval=60):
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/_sdk/_start_trace.py:64]: maybe we can leverage the heartbeat API to probe?

File: src/promptflow/promptflow/_sdk/_start_trace.py [hunk 3]
```diff
@@ -26,8 +31,10 @@ def start_trace():
 
     Note that this function is still under preview, and may change at any time.
     """
-    # TODO: PFS liveness probe
-    pfs_port = get_port_from_config()
+    pfs_port = get_port_from_config(create_if_not_exists=True)
+    _start_pfs_in_background(pfs_port)
+    # Start a liveness probe in a new thread
+    threading.Thread(target=_liveness_probe, args=(pfs_port,), daemon=True).start()
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_start_trace.py:37]: timeout for this action?

File: src/promptflow/promptflow/_sdk/_start_trace.py [hunk 4]
```diff
@@ -26,8 +31,10 @@ def start_trace():
 
     Note that this function is still under preview, and may change at any time.
     """
-    # TODO: PFS liveness probe
-    pfs_port = get_port_from_config()
+    pfs_port = get_port_from_config(create_if_not_exists=True)
+    _start_pfs_in_background(pfs_port)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_start_trace.py:35]: shall we check the pfs port alive before start it?

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 3 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py:37]: interval=60

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 4 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py:35]: yeah, have checked whether the port can be connected

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 3 above)
Comment (wangchao1230) [reply src/promptflow/promptflow/_sdk/_start_trace.py:37]: I think you mean the retry interval here. 
What I feels strange is why here is a while true probe here?

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 1 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py:54]: 1. update message here
2. add recording when set a pfs port and the port is not in use

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 2 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py:64]: moved

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 3 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py]: removed while true

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 2 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py]: updated according to offline sync

File: src/promptflow/promptflow/_sdk/_service/utils/utils.py [hunk 5]
```diff
@@ -45,6 +48,18 @@ def get_port_from_config(create_if_not_exists=False):
     return port
 
 
+def dump_port_to_config(port):
+    # Set port to ~/.promptflow/pf.yaml, if already have a port in file , will overwrite it.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/utils/utils.py:52]: the file is called pf.port.

File: src/promptflow/promptflow/_sdk/_start_trace.py [hunk 6]
```diff
@@ -41,6 +44,25 @@ def start_trace():
...
+    if is_port_in_use(pfs_port):
+        _logger.warning(f"Service port {pfs_port} is used.")
+        response = requests.get("http://localhost:{}/heartbeat".format(pfs_port))
+        if response.status_code != 200:
+            _logger.warning(f"Pfs service can't be reached through port {pfs_port}, will try to force restart pfs.")
+            args += ["--force"]
+        else:
+            _logger.warning(f"Pfs service is already running on port {pfs_port}, will not restart pfs.")
+            return
+    # Start a pfs process using detach mode
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/_start_trace.py:59]: Shall we wait until pfs's healthy here?

**Review by zhengfeiwang — APPROVED** [review]

File: src/promptflow/promptflow/_sdk/_start_trace.py [hunk 7]
```diff
@@ -50,34 +46,23 @@ def start_trace():
 
 def _start_pfs_in_background(pfs_port) -> None:
     """Start a pfs process in background."""
+    args = [sys.executable, "-m", "promptflow._sdk._service.entry", "start", "--port", str(pfs_port)]
     if is_port_in_use(pfs_port):
-        _logger.debug("PFS is already serving on port %s", pfs_port)
-        return
-    args = [sys.executable, "-m", "promptflow._sdk._service.entry", "start"]
+        _logger.warning(f"Service port {pfs_port} is used.")
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_start_trace.py:51]: we shoud have this logic extracted as function also

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 6 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py:58]: added

File: src/promptflow/promptflow/_sdk/_start_trace.py (hunk 7 above)
Comment (YingChen1996) [reply src/promptflow/promptflow/_sdk/_start_trace.py:53]: extracted

File: src/promptflow/promptflow/_trace/_start_trace.py [hunk 8]
```diff
@@ -44,22 +47,46 @@ def start_trace():
...
     if platform.system() == "Windows":
         os.spawnv(os.P_DETACH, sys.executable, args)
     else:
         os.system(" ".join(["nohup"] + args + ["&"]))
 
+    wait_time = 0
+    is_healthy = _check_pfs_service_status(pfs_port)
+    while is_healthy is False and time_threshold > wait_time:
+        _logger.info(f"Pfs service is not ready, will wait for at most {time_threshold}s.")
+        time.sleep(60)
```
[the remark this answers is not in the document]
Comment (wangchao1230) [reply src/promptflow/promptflow/_trace/_start_trace.py:69]: what's our avg start time?

[the remark this answers is not in the document]
Comment (YingChen1996) [reply src/promptflow/promptflow/_trace/_start_trace.py:69]: about 10s is ok, will update

File: src/promptflow/promptflow/_trace/_start_trace.py [hunk 9]
```diff
@@ -62,12 +62,16 @@ def _start_pfs_in_background(pfs_port) -> None:
     else:
         os.system(" ".join(["nohup"] + args + ["&"]))
 
-    wait_time = 0
+    wait_time = 10
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_trace/_start_trace.py]: We  should improve the time here and add a internal API telemetry. We can do it in next PR

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**